fix(network): Avoid initiating outbound handshakes with IPs for which Zebra already has an active peer.

arya2 opened the pull request:
## Motivation

This prevents some kinds of attacks.

Closes #6982.

## Solution

- Adds a `most_recent_by_ip` field to `AddressBook`.
- Updates the new field when there's an update with a later `last_response` time, or when addresses are removed.
- Adds `&& !self.has_active_peer_with_ip()` to the filter in `reconnection_peers()`. 
- Tests that the iterator from `reconnection_peers()` skips over address when there's already a live address with the same IP.
 
## Review

Part of regular scheduled work.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

- #7040

arya2 committed:
Adds most_recent_by_ip field to address book

arya2 committed:
adds test

arya2 commented on zebra-network/src/address_book.rs:
```diff
@@ -72,6 +73,9 @@ pub struct AddressBook {
     /// [`OrderedMap`] sorts in descending order.
     by_addr: OrderedMap<PeerSocketAddr, MetaAddr, Reverse<MetaAddr>>,
 
+    /// The most recently updated addresses by ip.
```
```suggestion
    /// The address with a last_connection_state of [`Responded`] and the most recent `last_response` time by IP.
```

arya2 commented on zebra-network/src/address_book.rs:
```diff
@@ -314,6 +325,33 @@ impl AddressBook {
+                    .map_or(true, |prev| {
+                        prev.last_response() < Some(updated_last_response)
+                    })
+            })
```
`last_response()` will currently only return `Some(_)` when `last_connection_state` is `Responded`, but in case that changes:

```suggestion
        updated.last_connection_state == PeerAddrState::Responded
            && updated
                .last_response()
                .is_some_and(|updated_last_response| {
                    self.most_recent_by_ip
                        .get(&updated.addr.ip())
                        .map_or(true, |prev| {
                            prev.last_response() < Some(updated_last_response)
                        })
                })
```

arya2 commented on zebra-network/src/address_book/tests/vectors.rs:
```diff
@@ -110,3 +110,38 @@ fn address_book_peer_order() {
+        DateTime32::MIN.saturating_add(Duration32::from_seconds(1)),
+    );
+
+    // Regardless of the order of insertion, the most recent address should be chosen first
```
```suggestion
    // The second address should be skipped because the first address has a
    // recent `last_response` time and the two addresses have the same IP.
```

arya2 committed:
Apply suggestions from code review

arya2 committed:
fixes lint

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -72,6 +73,10 @@ pub struct AddressBook {
     by_addr: OrderedMap<PeerSocketAddr, MetaAddr, Reverse<MetaAddr>>,
 
+    /// The address with a last_connection_state of [`PeerAddrState::Responded`] and
+    /// the most recent `last_response` time by IP.
```
`HashMap<IpAddr, MetaAddr>` only supports `max_conns_per_ip = 1`. That's ok for now, but needs to be documented here, on the constant, and in that user config field's docs.

Or we could mark that config as unsupported (developer/tester-only).

Supporting larger values is a different ticket with a different priority, that's not a security issue. And selfishly, I don't want to do that review right now, there's a lot of other things going on!

teor2345 replied:
If we're going to break support for initiating more than 1 outbound connection to an IP, we might want to delay merging this PR until after the next release. It's hard to say though, because it's only really important for testing.

What do you think users would want? ("They won't care" is an ok answer.)

arya2 replied:
> It's hard to say though, because it's only really important for testing.

I could open a suggestion PR replacing `has_active_peer_with_ip()` with `num_active_peers_with_ip()` to make the review a little easier if you think it may be worth doing for the tests?

> What do you think users would want?

I think users would expect `max_conns_per_ip` to at least configure the number of connections Zebra will accept, but that it would be better for it to define how many connections Zebra will try to initiate as well. It's not worth doing immediately unless we need it for integration tests.

teor2345 replied:
> > It's hard to say though, because it's only really important for testing.
> 
> I could open a suggestion PR replacing `has_active_peer_with_ip()` with `num_active_peers_with_ip()` to make the review a little easier if you think it may be worth doing for the tests?

That's a different change with a lower priority, because it's not a security fix. Also, I'd prefer we all shift to focusing on the RPC work soon, rather than adding more network changes. (I'd like to get our current set of changes tested in a release first.)

But ultimately this is @mpguerra's decision, and she uses tickets to make those decisions. 

> > What do you think users would want?
> 
> I think users would expect `max_conns_per_ip` to at least configure the number of connections Zebra will accept, but that it would be better for it to define how many connections Zebra will try to initiate as well. It's not worth doing immediately unless we need it for integration tests.

I think you're right, and it's ok as long as we document it.

arya2 replied:
Documented in https://github.com/ZcashFoundation/zebra/pull/7029/commits/d57f2b1c54883ae56ecec8f7fafba21a43663bc9.

teor2345 replied:
I was thinking about this a bit more yesterday. Some users will need Zebra to accept multiple connections from the same IP, either for testing, or for internal caching setups.

So could we disable the checks in this PR when `max_connections_per_ip > 1`?
(And document the security implications on that config.)

That way, the default config is secure, but users who need it can get Zebra to make multiple outbound connections.

arya2 replied:
Great idea, thank you, applied in https://github.com/ZcashFoundation/zebra/pull/7029/commits/6d17a9cbd0e37f2a5196d046e2c94dfc0eacd043.

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -397,6 +442,12 @@ impl AddressBook {
 
+                // Check if this surplus peer's addr matches that in `most_recent_by_ip`
+                // for this the surplus peer's ip to remove it there as well.
+                if self.is_addr_most_recent_by_ip(surplus_peer.addr) {
```
`AddressBook::take()` also needs to remove the IP from the cache. Are there any other methods that remove addresses from the address book?

arya2 replied:
> AddressBook::take() also needs to remove the IP from the cache.

Thank you.

> Are there any other methods that remove addresses from the address book?

No, the only other remove is in `get()` where it's immediately re-inserted.

teor2345 replied:
> > Are there any other methods that remove addresses from the address book?
> 
> No, the only other remove is in `get()` where it's immediately re-inserted.

Yes, as you can see I dislike that particular collection API 😂

arya2 replied:
Fixed in https://github.com/ZcashFoundation/zebra/pull/7029/commits/489e63d14a9c9db1796c275c60234f832b7f497a.

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -314,6 +326,34 @@ impl AddressBook {
+    /// - updated has a more recent last_response time than the existing [`MetaAddr`]
+    ///   for that IP in `most_recent_by_ip`.
+    fn should_update_most_recent_by_ip(&self, updated: MetaAddr) -> bool {
+        updated.last_connection_state == PeerAddrState::Responded
```
This condition implies an invariant, but I'm not sure exactly what it is.

Is it something like:
* the the only updates which change `last_response` are updates to the `Responded` state
* the only field checked by `was_recently_live()` is `last_response`

Are these invariants already documented, implemented, and tested somewhere else?

arya2 replied:
> Are these invariants already documented, implemented, and tested somewhere else?

It's the `MetaAddr` that's most likely to fail the `!MetaAddr.has_connection_recently_responded()` check in `MetaAddr.is_ready_for_connection_attempt()`. I'll add that to the method's docs.

The redundant `self.last_connection_state == PeerAddrState::Responded` is from `was_recently_live()` (though I'm not sure if it's there for clarity or to short-circuit).

teor2345 replied:
Hmm, I'm not sure if were talking about the same things here.

What I'm thinking of is documentation like:

> This code only needs to check a single cache entry, rather than the entire address book, because other code maintains these invariants:
> * the the only updates which change last_response are updates to the Responded state
> * the only field checked by was_recently_live() is last_response

arya2 replied:
Added in https://github.com/ZcashFoundation/zebra/pull/7029/commits/0c769b657fbe1f66dd20b7ed874ce4b529395395.

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -314,6 +326,34 @@ impl AddressBook {
+    /// - updated has a more recent last_response time than the existing [`MetaAddr`]
+    ///   for that IP in `most_recent_by_ip`.
+    fn should_update_most_recent_by_ip(&self, updated: MetaAddr) -> bool {
+        updated.last_connection_state == PeerAddrState::Responded
```
An IP that is currently in the `AttemptPending` state should also prevent other connections to that IP from being initiated. See my note below about how to implement this check.

What's the equivalent invariant for the `AttemptPending` state?

This will need a method docs update (but it doesn't need to be detailed, that's what the code and code comments are for).

arya2 replied:
What about `Failed`?

teor2345 replied:
An ~~IP~~ peer address and port that's currently in the `Failed` state can't have an outbound connection:

https://github.com/ZcashFoundation/zebra/blob/31bc46bb2a736110d215de2fac5cdfd79bcbd058/zebra-network/src/meta_addr.rs#L66-L68

Edit: it's based on address and port, not just IP

arya2 replied:
> it's based on address and port, not just IP

Should it check for recent failures by IP as well?

teor2345 replied:
For security reasons, I think failures should be ignored entirely in this security fix:
https://github.com/ZcashFoundation/zebra/pull/7029#discussion_r1252379827

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -314,6 +326,34 @@ impl AddressBook {
+    }
+
+    /// Returns true if the provided `addr` matches that in `most_recent_by_ip`
+    /// for the IP of the address.
```
Nitpick: this doc is a bit unclear to me. I had to read the code to find out it was an exact IP and port match, rather than just an IP address match. ("Address" can mean either.)

It's tricky to explain, here's my attempt at it:
```suggestion
    /// Returns true if `addr` is the latest entry for its IP, which is stored in `most_recent_by_ip`.
    /// The entry is checked for an exact match to the IP and port of `addr`.
```

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -314,6 +326,34 @@ impl AddressBook {
         meta_addr
     }
 
+    /// Returns true if `updated` has a last_connection_state of `Responded`
```
Nitpick: Can we add a summary line to explain what the method does?

```suggestion
    /// Returns true if `updated` needs to be applied to the recent outbound peer connection IP cache.
    ///
    /// Cache entries are updated if they have a last_connection_state of `Responded`
```

arya2 replied:
Applied in https://github.com/ZcashFoundation/zebra/pull/7029/commits/77833d9ab6eac162f5e0e39b68d35187ab33cc63.

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -463,6 +514,13 @@ impl AddressBook {
+    fn has_active_peer_with_ip(&self, ip: &IpAddr, chrono_now: chrono::DateTime<Utc>) -> bool {
+        self.most_recent_by_ip
+            .get(ip)
+            .map_or(false, |peer| peer.was_recently_live(chrono_now))
```
We also need to check `peer.was_connection_recently_attempted()` here, and update the method docs.

arya2 replied:
Updated to check last_attempt, last_response, and last_failure times in https://github.com/ZcashFoundation/zebra/pull/7029/commits/489e63d14a9c9db1796c275c60234f832b7f497a.

teor2345 replied:
Why does last_failure matter?

arya2 replied:
So that `reconnection_peers()` skips addresses with IPs where there's a recently failed connection as it would if the connection had recently failed (in `is_ready_for_connection_attempt()`).

> Why does last_failure matter?

I don't think it actually does matter, it's just what I would expect it to be doing if I were debugging later.

teor2345 replied:
I think that checking the failure time enables an attack which pollutes our address book with fake peers on the same IP as a real peer. If the real peer is on a non-default port, the fake peers will all be tried before it, delaying a connection to the real peer by 2 minutes per fake peer. (Which is effectively never, because each IP address has 65000 TCP ports, so the limit is the 4000 address book addresses.)

So I think we should treat recently successful connections differently to connections that were never successful and immediately failed.

Specifically, an outbound connection to the same IP should not be attempted when:
* the peer recently had a successful connection and responded (~ 2 minute timeout)

But an outbound connection to the same IP should be attempted even if:
* we recently attempted a handshake to that peer
* we recently failed a connection to that peer without completing the handshake (`last_response` was not updated)

If we want to do anything about avoiding attempting multiple connections to the same IP at the same time, let's do that in a separate ticket with a lower priority?

arya2 replied:
> But an outbound connection to the same IP should be attempted even if:
> 
> - we recently attempted a handshake to that peer
> - we recently failed a connection to that peer without completing the handshake (last_response was not updated)

Fixed in https://github.com/ZcashFoundation/zebra/pull/7029/commits/788bce3b9ea89d3d5b3d535d09720b9b04c169a0 and https://github.com/ZcashFoundation/zebra/pull/7029/commits/f8338f1871f829851582fd3d7015effa7f0e04be.

teor2345 commented on zebra-network/src/address_book/tests/vectors.rs:
```diff
@@ -110,3 +110,39 @@ fn address_book_peer_order() {
+/// Check that `reconnection_peers` skips addresses with IPs for which
+/// Zebra already has recently connected peers.
+#[test]
+fn reconnection_peers_skips_live_ip() {
```
Can we have a similar test with AttemptPending, and one where the cache entry was recently Attempted/Responded, but has now Failed?

arya2 replied:
Added in https://github.com/ZcashFoundation/zebra/pull/7029/commits/77833d9ab6eac162f5e0e39b68d35187ab33cc63.

teor2345 replied:
I'm having trouble working out what the tests do, and what the closure argument means. Would you mind documenting what each test case covers?

arya2 replied:
Good point, added comments in https://github.com/ZcashFoundation/zebra/pull/7029/commits/9897b37cc98faa8ff8c0241515d4f0150a6f59fe.

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -314,6 +326,34 @@ impl AddressBook {
+    /// - updated has a more recent last_response time than the existing [`MetaAddr`]
+    ///   for that IP in `most_recent_by_ip`.
+    fn should_update_most_recent_by_ip(&self, updated: MetaAddr) -> bool {
+        updated.last_connection_state == PeerAddrState::Responded
```
Should moving that `MetaAddr.addr` to a `Failed` state invalidate the cached entry, and replace it with the next oldest entry for that IP?

Or is there some invariant that makes it ok to keep the latest entry, even if its state has changed since?

I think the invariant is something like:
* `last_response` and `last_attempt` times only increase

But I can't remember if that holds after our latest fixes.

arya2 replied:
> Should moving that MetaAddr.addr to a Failed state invalidate the cached entry, and replace it with the next oldest entry for that IP?

I don't think so, it should always prefer the `meta_addr` that will fail `is_ready_for_connection_attempt()` the longest, including `last_failure` times, and it should check all of the times that `is_ready_for_connection_attempt()` checks in a renamed `has_active_peer_with_ip()`.

> last_response and last_attempt times only increase

That's still the case.

arya2 replied:
Documented in https://github.com/ZcashFoundation/zebra/pull/7029/commits/0c769b657fbe1f66dd20b7ed874ce4b529395395.

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -463,6 +514,13 @@ impl AddressBook {
+    fn has_active_peer_with_ip(&self, ip: &IpAddr, chrono_now: chrono::DateTime<Utc>) -> bool {
+        self.most_recent_by_ip
+            .get(ip)
+            .map_or(false, |peer| peer.was_recently_live(chrono_now))
```
Note: `last_response` times can remain live for a long time if the local clock is changed to a much earlier time. There's not much we can do about that, but it might be worth documenting, since checking by IP makes the problem slightly worse.

(It impacts all nodes with that IP, not just the node with that exact IP and port.)

arya2 replied:
Documented in https://github.com/ZcashFoundation/zebra/pull/7029/commits/0c769b657fbe1f66dd20b7ed874ce4b529395395.

arya2 replied:
Should it be updated to make `last_response` an `Instant`, or a tuple with both?

teor2345 replied:
> Should it be updated to make `last_response` an `Instant`, or a tuple with both?

I think that's beyond the scope of a small security fix?

(We need to have a clock time because responded times are sent as part of the Zcash network protocol.)

teor2345 reviewed (changes requested):
This is a nice minimal design!

I think checking `AttemptPending` peers is a blocking change. If we decide to put this PR in a later release, that could also be blocking.

The rest of my comments provide context, ask for more context so we can check the design works as expected, or ask for docs so future developers and auditors can do those checks.

Feel free to ask me to find some of these details myself.

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -463,6 +514,13 @@ impl AddressBook {
         self.by_addr.descending_values().cloned()
     }
 
+    /// Returns true if there is a recently live peer connection with the provided IP address.
```
```suggestion
    /// Returns true if there is a recently live outbound peer connection with the provided IP address.
```

arya2 replied:
Applied in https://github.com/ZcashFoundation/zebra/pull/7029/commits/0c769b657fbe1f66dd20b7ed874ce4b529395395.

teor2345 commented on zebra-network/src/address_book/tests/vectors.rs:
```diff
@@ -110,3 +110,39 @@ fn address_book_peer_order() {
 }
+
+/// Check that `reconnection_peers` skips addresses with IPs for which
+/// Zebra already has recently connected peers.
```
```suggestion
/// Zebra already has recently connected outbound peers.
```

arya2 replied:
It looks like the address book should be tracking inbound connections as well: https://github.com/ZcashFoundation/zebra/blob/issue6681/zebra-network/src/peer/handshake.rs#L933

teor2345 replied:
This method returns `None` for inbound connections, so the address book never finds out about them:
https://github.com/ZcashFoundation/zebra/blob/b6ba3f70319f8283feb8e7033b25f453b90ad843/zebra-network/src/peer/handshake.rs#L245-L254
https://github.com/ZcashFoundation/zebra/blob/b6ba3f70319f8283feb8e7033b25f453b90ad843/zebra-network/src/peer/handshake.rs#L263-L270

arya2 replied:
Applied in https://github.com/ZcashFoundation/zebra/pull/7029/commits/0c769b657fbe1f66dd20b7ed874ce4b529395395.

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -3,8 +3,9 @@
 
```
The title of this PR needs a tweak, because the `AddressBook` only stores the addresses of outbound peer connections.

So this PR only prevents rapidly repeated outbound reconnections and handshakes to the same IP. (Which is the attack I'm trying to address, and what I should have written in the ticket.)

It's ok to leave any duplicate outbound/inbound IPs for the peer set to drop. That would be a different attack, which seems much less plausible to me. But if you can think of a high-impact attack across inbound and outbound connections, please open a ticket!

arya2 committed:
Updates most_recent_by_ip in .take()

Updates should_update_most_recent_by_ip() and has_active_peer_with_ip to check last_attempt and last_failure times

Renames has_active_peer_with_ip

arya2 commented on zebra-network/src/address_book.rs:
```diff
@@ -463,6 +521,18 @@ impl AddressBook {
+        instant_now: Instant,
+        chrono_now: chrono::DateTime<Utc>,
+    ) -> bool {
+        self.most_recent_by_ip.get(ip).map_or(false, |peer| {
```
```suggestion
        self.most_recent_by_ip.get(ip).map_or(true, |peer| {
```

arya2 committed:
Documents that Zebra will not initiate more than 1 outbound connections per IP

arya2 committed:
Fixes is_ready_for_connection_attempt_with_ip

Adds test coverage for AttemptPending and Failed

Fixes new_outbound_peer_connections_are_rate_limited proptest

teor2345 commented:
I have about 5 things on my TODO list before I get to this review, so I'm not sure if it's going to make it into this release. You could ask Pili to give the review to someone else, or re-prioritise this ticket?

arya2 committed:
Applies suggestions from code review.

arya2 committed:
Applies suggestions from code review

arya2 committed:
Always return true from `is_ready_for_connection_attempt_with_ip` if max_connections_per_ip != 0

arya2 committed:
Update max_connections_per_ip config docs

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -141,6 +155,8 @@ impl AddressBook {
         let (address_metrics_tx, _address_metrics_rx) = watch::channel(AddressMetrics::default());
 
+        // Avoid initiating outbound handshakes when max_connections_per_ip is 1.
+        let should_limit_outbound_conns_per_ip = max_connections_per_ip == 1;
```
Currently we assert that `initial_peer_set_size` is greater than zero, should we do the same with `max_connections_per_ip`?

Or should we remove the assertions on both and just warn?

```suggestion
        let should_limit_outbound_conns_per_ip = max_connections_per_ip <= 1;
```

arya2 replied:
I think we should remove the assertions on both, use the default values, and log a warning.

Update: added warnings in https://github.com/ZcashFoundation/zebra/pull/7029/commits/3ac16d57cafd380f5678f0a1efc373bdfccb7b1a.

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -72,6 +73,14 @@ pub struct AddressBook {
+    ///
+    /// This is used to avoid initiating outbound connections past [`Config::max_connections_per_ip`](crate::config::Config), and
+    /// currently only supports a `max_connections_per_ip` of 1, and must be `None` when used with a greater `max_connections_per_ip`.
+    // TODO: Replace with `by_ip` to support configured `max_connections_per_ip`
```
Optional:

This is a potential design for another ticket, which would keep the computational complexity to `O(1)`, except for removal which would `O(addresses for that IP with the same last responded time)`. (Which is usually 1.)

```suggestion
    // TODO: Replace with `by_ip: HashMap<IpAddr, BTreeMap<DateTime32, MetaAddr>>` to support configured `max_connections_per_ip` greater than 1
```

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -314,6 +340,49 @@ impl AddressBook {
+    /// This code only needs to check a single cache entry, rather than the entire address book,
+    /// because other code maintains these invariants:
+    /// - `last_attempt`, `last_failure`, and `last_response` times for an entry can only increase.
+    /// - these are the only fields checked by `was_recently_updated()`
```
If we just look at last response, this function (and its docs) needs to be updated.

```suggestion
    /// - `last_response` time for an entry can only increase
    /// - this is the only field checked by `has_connection_recently_responded()`
```

arya2 replied:
Applied in https://github.com/ZcashFoundation/zebra/pull/7029/commits/788bce3b9ea89d3d5b3d535d09720b9b04c169a0.

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -463,6 +556,23 @@ impl AddressBook {
     }
 
+    /// Is this IP ready for a new outbound connection attempt?
+    /// Checks if the most recently updated outbound address with this IP `was_recently_updated()`.
```
If we just look at last response, this function (and its docs) needs to be updated.



```suggestion
    /// Checks if the most recently updated outbound address with this IP `has_connection_recently_responded()`.
```

arya2 replied:
Applied something similar in https://github.com/ZcashFoundation/zebra/pull/7029/commits/788bce3b9ea89d3d5b3d535d09720b9b04c169a0.

teor2345 commented on zebra-network/src/constants.rs:
```diff
@@ -71,6 +71,12 @@ pub const OUTBOUND_PEER_LIMIT_MULTIPLIER: usize = 3;
+///       has already recently sent or received a message to or from an address
+///       with that IP. Zebra will not initiate more than 1 outbound connection
+///       to an IP based on this configuration, but it will accept more inbound
+///       connections to an IP.
```
If we just look at last response, these docs need to be updated.

```suggestion
/// Note: Zebra will currently avoid initiating outbound connections where it
///       has recently had a successful handshake with any address
///       on that IP. Zebra will not initiate more than 1 outbound connection
///       to an IP based on the default configuration, but it will accept more inbound
///       connections to an IP.
```

teor2345 commented on zebra-network/src/config.rs:
```diff
@@ -158,6 +158,12 @@ pub struct Config {
+    ///       additional outbound handshakes. It will otherwise still initiate outbound handshakes
+    ///       beyond this limit. Zebra will also currently still accept more inbound connections
+    ///       per IP. In cases where additional handshakes are permitted beyond this limit,
+    ///       connections will be dropped when attempting to add them to the peer set.
```
For security-sensitive configs, it's important to warn the user in the first sentence.

Can I suggest a format for user-focused documentation?
- Short summary of the most important user impact
- Summary of how to avoid that impact
- Technical details

```suggestion
    /// # Security
    ///
    /// Increasing this config above 1 reduces Zebra's network security.
    ///
    /// If this config is greater than 1, Zebra can initiate multiple outbound handshakes to the same
    /// IP address.
    ///
    /// This config does not currently limit the number of inbound connections that Zebra will accept
    /// from the same IP address.
    ///
    /// If Zebra makes multiple inbound or outbound connections to the same IP, they will be dropped
    /// after the handshake, but before adding them to the peer set. The total numbers of inbound and
    /// outbound connections are also limited to a multiple of `peerset_initial_target_size`.
```

teor2345 reviewed (changes requested):
It seems like the scope of this PR has grown a bit. I'm partly responsible for that, I think my suggestion about pending handshakes was out of scope. (And it's difficult to do securely.)

So I'd like to make sure all the code we add in this PR has a strong security motivation. That way, it's less likely to have other negative security impacts.

arya2 committed:
Warn about invalid config fields and use default values

arya2 committed:
Ignores last_attempt and last_failure in is_ready_for_connection_attempt_with_ip

updates test

arya2 committed:
Only update most_recent_by_ip if update.last_conn_state is responded.

arya2 committed:
Apply suggestions from code review

Co-authored-by: teor <teor@riseup.net>

arya2 committed:
fixes lint

arya2 committed:
Update zebra-network/src/address_book.rs

Co-authored-by: teor <teor@riseup.net>

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -435,6 +518,12 @@ impl AddressBook {
+            // Check if this surplus peer's addr matches that in `most_recent_by_ip`
+            // for this the surplus peer's ip to remove it there as well.
+            if self.should_remove_most_recent_by_ip(entry.addr) {
+                self.most_recent_by_ip.as_mut()?.remove(&entry.addr.ip());
```
I think this will return `None` when `most_recent_by_ip` is disabled, even if a peer was actually removed. But this method should return the removed peer.

```suggestion
                if let Some(most_recent_by_ip) = self.most_recent_by_ip.as_mut() {
                    most_recent_by_ip.remove(&entry.addr.ip());
                }
```

arya2 replied:
> I think this will return `None`

It wasn't clear but it's guaranteed to be `Some` if `should_remove_most_recent_by_ip()` is true.

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -463,6 +552,22 @@ impl AddressBook {
+            .as_ref()
+            .and_then(|most_recent_by_ip| most_recent_by_ip.get(ip))
+            .filter(|peer| peer.has_connection_recently_responded(chrono_now))
+            .is_none()
```
I think the logic here might be inverted?

But I'm not sure, because the tests look like they are correct, and they are passing. Maybe we're missing a test for this specific case.

Here's a minimal fix:

```rust
            .filter(|peer| !peer.has_connection_recently_responded(chrono_now))
```

Here's a refactor where that is easier to see:
```suggestion
        let Some(most_recent_by_ip) = self.most_recent_by_ip.as_ref() else {
            // if we're not checking IPs, any connection is allowed
            return true;
        }
        
        let Some(same_ip_peer) = most_recent_by_ip.get(ip) else {
            // If there's no entry for this IP, any connection is allowed
            return true;
        }
        
        !same_ip_peer.has_connection_recently_responded(chrono_now)
```

arya2 replied:
> I think the logic here might be inverted?
> 
> But I'm not sure

The logic was correct, but your version is much clearer.

teor2345 reviewed (dismissed):
Looks great!

I just want to double-check the logic in two places where we might be returning an incorrect value.

arya2 committed:
Apply suggestions from code review

Co-authored-by: teor <teor@riseup.net>

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -463,6 +554,28 @@ impl AddressBook {
+        let Some(most_recent_by_ip) = self.most_recent_by_ip.as_ref() else {
+            // if we're not checking IPs, any connection is allowed
+            return true;
+        }
```
oops

```suggestion
        };
```

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -463,6 +554,28 @@ impl AddressBook {
+        let Some(same_ip_peer) = most_recent_by_ip.get(ip) else {
+            // If there's no entry for this IP, any connection is allowed
+            return true;
+        }
```
```suggestion
        };
```

teor2345 committed:
Fix Rust syntax

teor2345 reviewed (dismissed):
Sorry about the Rust syntax, I didn't check it before suggesting

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -463,6 +554,28 @@ impl AddressBook {
+            // if we're not checking IPs, any connection is allowed
+            return true;
+        };
+        
```
```suggestion

```

teor2345 commented on zebra-network/src/address_book.rs:
```diff
@@ -463,6 +554,28 @@ impl AddressBook {
+            // If there's no entry for this IP, any connection is allowed
+            return true;
+        };
+        
```
```suggestion

```

teor2345 committed:
Fix whitespace

arya2 commented:
@Mergifyio refresh
